tester/src/test/scala/spinal/tester/scalatest/FormalFifoTester.scala

```diff
@@ -44,23 +61,58 @@ class FormalFifoTester extends SpinalFormalFunSuite {
         val d1 = anyconst(UInt(7 bits))
         val d2 = anyconst(UInt(7 bits))
 
+        // assume first d1, then d2 enters the StreamFifo
         val (d1_in, d2_in) = dut.io.push.formalAssumesOrder(d1, d2)
+        // assert first d1, then d2 leaves the StreamFifo
         val (d1_out, d2_out) = dut.io.pop.formalAssertsOrder(d1, d2)
 
+        // make sure d1 is not already in the StreamFifo before it enters
         when(!d1_in) { assume(!dut.formalContains(d1)) }
+        // make sure d1 only exists once inside the StreamFifo after it entered and before it left
         when(d1_in && !d1_out) { assert(dut.formalCount(d1) === 1) }
 
         when(!d2_in) { assume(!dut.formalContains(d2)) }
         when(d2_in && !d2_out) { assert(dut.formalCount(d2) === 1) }
 
+        // make sure that d2 does not leave before d1 leaves the StreamFifo
         when(d1_in && d2_in && !d1_out) { assert(!d2_out) }
 
+        // cover the case that FIFO first goes to full, then to empty
+        dut.formalFullToEmpty()
+
+        // TODO for .withProve() induction prove to pass, relationship between push/pop distance and some fill rate
```

Comment:
are the TODOs in here (and the one below on line 124) still a thing?

Reply:
Yes, the TODO's are up-to-date with the code.

I have newer formal code with update TODO's in my `fifo-rework` branch fork, matching better formal coverage.